### PR TITLE
feat: add page for describing cairo builtins

### LINKED ISSUE #1126: Request for Contribution: Describe Cairo builtins
# Description
We need a page that describes what Cairo builtins are, what they do generally, and a list of builtins with a short description of what each one does specifically.

Explain what high-level Cairo code ends up invoking a given builtin? For example:

* `a&b` is a single application of the bitwise builtin. 
* `a>b` is a rangecheck.
# Source material

* The [builins page](https://docs.cairo-lang.org/how_cairo_works/builtins.html) from the CairoZero documentation.
* Gas and transaction fees>Fee calculation>Calculation of computation costs>[Table 1. Amount of gas used per Cairo step or per each time a Cairo builtin is applied](https://docs.starknet.io/documentation/architecture_and_concepts/Network_Architecture/fee-mechanism/#gas_cost_per_cairo_step_or_builtin_step)
* The files in this directory might also be useful: https://github.com/lambdaclass/cairo-vm/blob/352e8a8cb8addc267d8ae59073b6447fd8517c81/vm/src/vm/runners/builtin_runner
# Instructions for Contributing
1. Leave a comment on this Issue indicating that you want to take ownership of this feature.
2. Wait for our confirmation to start on the project.
3. Once you have received confirmation, begin working on your contribution in a new branch.
4. Create a new AsciiDoc file named `builtins.adoc` in `components/Starknet/modules/architecture_and_concepts/pages/Smart_Contracts`. 
5. Add the new file to the table of contents by adding an `xref` to it in `components/Starknet/modules/architecture_and_concepts/nav.adoc`
6. IMPORTANT: Do not simply copy and paste existing text in the source material above. Follow these [basic writing guidelines](https://github.com/starknet-io/starknet-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#basic-writing-guidelines). Some basic guidelines include: Use the active voice, not the passive voice, talk to your reader directly by using the second-person (you) and imperative. Use the present tense, avoid the future tense.
7. Be ready to make changes based on review comments.
# Remuneration
Compensation for the approved contributions will be provided via OnlyDust.
# Deadline
Maximum of 10 days
# Contact
If you have any questions or need further details, please contact @stoobie.
_____
We look forward to your proposals. Thank you for contributing to the Starknet ecosystem!

### PR DESCRIPTION
**Pull request by xiaolou86**
### Description of the Changes

Closes #1126  
Add pages for describing cairo builtins

### PR Preview URL

After you push a commit to this PR, a preview is built and a URL to the root of the preview appears in the comment feed.

Paste here the specific URL(s) of the content that this PR addresses.

### Check List

- [x] Changes have been done against main branch, and PR does not conflict
- [x] PR title follows the convention: `<docs/feat/fix/chore>(optional scope): <description>`, e.g: `fix: minor typos in code`

- - -
This change is [Reviewable](https://reviewable.io/reviews/starknet-io/starknet-docs/1202)

### CODE REVIEW

**Comment by xiaolou86** [comment]
@stoobie I finished the doc, please help me to review. Thank you very much.

**Comment by stoobie** [comment]
@xiaolou86 Thank you! I've forwarded this on for a technical review. Once that is done, I'll go over it and edit or make suggestions.

**Comment by xiaolou86** [comment]
ok, I will update the docs.

**Comment by xiaolou86** [comment]
@stoobie please help to review again.

**Comment by stoobie** [comment]
Please mention that you don't need to do anything to _use_ a builtin. The library you use should automatically scan your code and use a builtin where appropriate. I'm not 100% sure if I got this right, so please wait for a confirmation from @ArielElp.

Also, the only builtins that you should have on this page are the ones [isted in on the docs list in the fee page in [Table 1. Amount of gas used per Cairo step or per each time a Cairo builtin is applied](https://docs.starknet.io/documentation/architecture_and_concepts/Network_Architecture/fee-mechanism/#gas_cost_per_cairo_step_or_builtin_step))

File: components/Starknet/modules/architecture_and_concepts/pages/Smart_Contracts/builtins.adoc
Comment (stoobie) [inline components/Starknet/modules/architecture_and_concepts/pages/Smart_Contracts/builtins.adoc:3]: I don't think you actually need headings. This section is so small, and you can just use a table title instead of a heading for the table.

File: components/Starknet/modules/architecture_and_concepts/pages/Smart_Contracts/builtins.adoc
Comment (stoobie) [inline components/Starknet/modules/architecture_and_concepts/pages/Smart_Contracts/builtins.adoc:5]:
```suggestion
_Builtins_, in Cairo, are predefined optimized low-level execution units that the Cairo VM uses to perform predefined computations that are expensive to perform in standard Cairo. Builtins enhance the functionality of the Cairo VM, enabling you to perform tasks such as using the Poseidon hash, range-checks, or ECDSA signature verifications, efficiently and using less gas.
```
* _Allow_- means that you are not prevented from doing something. In contrast, _enable_ means that you can do something that you couldn't have done otherwise.
* _which_ vs _that_: In American English, which is the industry standard in technical documentation (as opposed to British English), _which_ sets off a part of a sentence that you remove entirely and the sentence would still make sense. A clause beginning with _that_ is an intrinsic part of the sentence.

**Review by stoobie — CHANGES_REQUESTED** [review]
@xiaolou86 I made a couple more comments/suggestions. I asked someone more knowledgeable than me to comment as well, so please be aware that he needs to respond. You can make the minor changes I requested in the meantime.

File: components/Starknet/modules/architecture_and_concepts/pages/Smart_Contracts/builtins.adoc
Comment (stoobie) [inline components/Starknet/modules/architecture_and_concepts/pages/Smart_Contracts/builtins.adoc:7]:
```suggestion
[#list-of-cairo-builtins]
.List of Cairo builtins
```

**Comment by xiaolou86** [comment]
> Please mention that you don't need to do anything to _use_ a builtin. The library you use should automatically scan your code and use a builtin where appropriate. I'm not 100% sure if I got this right, so please wait for a confirmation from @ArielElp.

All right.

> 
> Also, the only builtins that you should have on this page are the ones [isted in on the docs list in the fee page in [Table 1. Amount of gas used per Cairo step or per each time a Cairo builtin is applied](https://docs.starknet.io/documentation/architecture_and_concepts/Network_Architecture/fee-mechanism/#gas_cost_per_cairo_step_or_builtin_step))

I think I have the only ones on this page, but I am not much sure. 

Thank you very much.

File: components/Starknet/modules/architecture_and_concepts/pages/Smart_Contracts/builtins.adoc
Comment (stoobie) [inline components/Starknet/modules/architecture_and_concepts/pages/Smart_Contracts/builtins.adoc:4]:
```suggestion
_Builtins_, in Cairo, are predefined optimized low-level execution units that the Cairo VM uses to perform predefined computations that are expensive to perform in standard Cairo. Builtins enhance the functionality of the Cairo VM, enabling you to perform certain tasks, such as using the Poseidon hash, range-checks, or ECDSA signature verifications, more efficiently, using less gas.
```

**Comment by stoobie** [comment]
@xiaolou86 After reviewing this with our expert, I found that we were on the wrong track. I created a new PR (#1243) with the necessary info.

Feel free to review it and make suggestions or ask questions if anything is unclear. After the new PR is merged, I'll settle with you on OnlyDust.

**Comment by stoobie** [comment]
#1243 Closes this.

**Pull request closed without merging**
